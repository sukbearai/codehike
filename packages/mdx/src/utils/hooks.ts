import React from "react"

export const useLayoutEffect =
  typeof window !== "undefined"
    ? React.useLayoutEffect
    : React.useEffect

// for debugging:
// export const useLayoutEffect = (
//   effect: any,
//   deps?: any
// ) => {}

// from https://overreacted.io/making-setinterval-declarative-with-react-hooks/
export function useInterval(
  callback: () => void,
  delay: number | null
) {
  const savedCallback = React.useRef(callback)

  React.useEffect(() => {
    savedCallback.current = callback
  }, [callback])

  React.useEffect(() => {
    if (!delay && delay !== 0) {
      return undefined
    }

    const id = setInterval(
      () => savedCallback.current(),
      delay
    )

    return () => clearInterval(id)
  }, [delay])
}
