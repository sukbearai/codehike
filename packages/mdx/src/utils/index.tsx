export { transparent } from "./color"
export * from "./tween"
export * from "./code"
export * from "./focus"
export * from "./hooks"

export function clamp(x: number, min: number, max: number) {
  return Math.min(Math.max(x, min), max)
}
