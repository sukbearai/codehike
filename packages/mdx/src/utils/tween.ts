export type Tween<T> =
  | { prev?: T; next: T }
  | { prev: T; next?: T }

export type FullTween<T> = { prev: T; next: T }

type AnyTween<T> = Tween<T> | FullTween<T>

function map<T, R>(
  tween: FullTween<T>,
  fn: (t: T, key: "prev" | "next") => R
): FullTween<R>
function map<T, R>(
  tween: Tween<T>,
  fn: (t: T | undefined, key: "prev" | "next") => R
): Tween<R>
function map<T, R>(
  tween: AnyTween<T>,
  fn: (t: T | undefined, key: "prev" | "next") => R
): AnyTween<R> {
  return {
    prev: fn(tween.prev, "prev"),
    next: fn(tween.next, "next"),
  }
}

export { map }

export function withDefault<T>(
  t: Tween<T> | undefined,
  deft: T
): FullTween<T> {
  return {
    prev: t?.prev === undefined ? deft : t.prev,
    next: t?.next === undefined ? deft : t.next,
  }
}

export function mapWithDefault<T, R>(
  tween: Tween<T> | undefined,
  deft: T,
  fn: (t: T, key: "prev" | "next") => R
): FullTween<R> {
  return map(withDefault(tween, deft), fn)
}

export function anyValue<T, R>(
  tween: AnyTween<T>,
  fn: (t: T) => R
): R {
  return fn(tween.prev) || fn(tween.next)
}
