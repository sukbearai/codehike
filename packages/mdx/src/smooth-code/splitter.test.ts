import { splitTokens } from "./splitter"
import { parseExtremes } from "../utils"
import { describe, it, expect, test } from "vitest"

describe("split tokens", () => {
  it("with no extremes", () => {
    const tokens = splitTokens(asTokens("foo", "bar"), [])
    expect(tokens).toMatchSnapshot()
  })
  it("with extremes", () => {
    const tokens = splitTokens(
      asTokens("foo", "bar"),
      asExtremes("1:2", "6")
    )
    expect(tokens).toMatchSnapshot()
  })
})

function asTokens(...strings: string[]) {
  return strings.map(content => ({ content, props: {} }))
}

function asExtremes(...strings: string[]) {
  return strings.map(s => parseExtremes(s))
}
