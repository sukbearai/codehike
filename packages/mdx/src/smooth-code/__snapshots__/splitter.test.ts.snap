// Vitest Snapshot v1

exports[`split tokens > with extremes 1`] = `
[
  {
    "content": "fo",
    "props": {},
  },
  {
    "content": "o",
    "props": {},
  },
  {
    "content": "ba",
    "props": {},
  },
  {
    "content": "r",
    "props": {},
  },
]
`;

exports[`split tokens > with no extremes 1`] = `
[
  {
    "content": "foo",
    "props": {},
  },
  {
    "content": "bar",
    "props": {},
  },
]
`;
